src/compiler/checker.ts

```diff
@@ -19193,7 +19194,7 @@ namespace ts {
                 getContextualTypeForObjectLiteralMethod(node, contextFlags) :
                 getContextualType(node, contextFlags);
             const instantiatedType = instantiateContextualType(contextualType, node, contextFlags);
-            if (instantiatedType) {
+            if (instantiatedType && !(contextFlags && contextFlags & ContextFlags.NoConstraints && instantiatedType.flags & TypeFlags.TypeVariable)) {

```

Comment:
In addition to `TypeVariable`, should this also be looking for `Substitution` (since that wraps a type variable pretty directly)?

Reply:
Can you write an example that would depend on this?

Reply:
```ts
declare function foo<T extends (x: string) => string>(f: T): T;

type SubstituteOf<T> = T extends {} ? T : never;

const f = foo(<U>(x: SubstituteOf<U>) => x);  // <U>(x: U) => U
```
still instantiates as `string` on this PR.

Reply:
No, for your example you get what you'd expect with the PR:

```ts
const f = foo(<U>(x: SubstituteOf<U>) => x);  // <U>(x: SubstituteOf<U>) => SubstituteOf<U>
```

The substitution type would need to be in the contextual type (i.e. the parameter's type would need to be a substitution type) for the original question you posed to make sense. I don't see how that could happen which is why I asked for an example.